pkg/federate/kubefed/cluster_test.go

```diff
@@ -0,0 +1,252 @@
+package kubefed
+
+import (
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+	fedv1 "sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+)
+
+type added string
+type updated string
+type removed string
+
+type testClusterEventHandler struct {
+	id     string
+	events chan interface{}
+}
+
+const (
+	secretName = "my-secret"
+)
+
+var _ = Describe("Kubefed Cluster", func() {
+	klog.InitFlags(nil)
+
+	Describe("Adding a watch", testAddingWatch)
+	Describe("When KubeFedCluster items are added, updated, removed", testOnKubeFedClusterChanges)
+	Describe("clusterWatcher", testClusterWatcher)
+})
+
+func testAddingWatch() {
+	var stopChan chan struct{}
+	BeforeEach(func() {
+		stopChan = make(chan struct{})
+	})
+
+	AfterEach(func() {
+		close(stopChan)
+	})
+
+	It("Should notify ClusterEventHandler of existing KubeFedCluster items", func() {
+		federator := newFederator(newInformer(watch.NewFake(), *newKubeFedCluster("east"),
+			*newKubeFedCluster("west")), stopChan)
+
+		handler1 := newtTestClusterEventHandler("1")
+		err := federator.WatchClusters(handler1)
+		Expect(err).ToNot(HaveOccurred())
+
+		Eventually(handler1.events, 5).Should(Receive(Or(Equal(added("east")), Equal(added("west")))))
+		Eventually(handler1.events, 5).Should(Receive(Or(Equal(added("east")), Equal(added("west")))))
+		Consistently(handler1.events).ShouldNot(Receive())
+
+		handler2 := newtTestClusterEventHandler("2")
+		err = federator.WatchClusters(handler2)
+		Expect(err).ToNot(HaveOccurred())
+
+		Eventually(handler2.events, 5).Should(Receive(Or(Equal(added("east")), Equal(added("west")))))
+		Eventually(handler2.events, 5).Should(Receive(Or(Equal(added("east")), Equal(added("west")))))
+	})
+}
+
+func testOnKubeFedClusterChanges() {
+	var stopChan chan struct{}
+	BeforeEach(func() {
+		stopChan = make(chan struct{})
+	})
+
+	AfterEach(func() {
+		close(stopChan)
+	})
+
+	It("Should notify ClusterEventHandler of appropriate change events", func() {
```

Comment:
the By() is a nice addition they have but it would be nicer to divide this test in smaller Its perhaps inside a When() or Context() 

Is there an advantage of doing it this way? I mean, are you testing something implicitly by doing it like this?

Reply:
The intention here is to test scenarios in sequence that rely on previous steps and verifies correctness of internal state tracking. Separating into multiple `When()/Context()` blocks would not achieve that.

Reply:
yeah, I understand that you are implicitly testing that. I just favor more and divided tests that check the things explicitly and that give more real coverage of the use cases we are building in the code. 

Ideally tests should be short, concise and clear. That provides the best behaviour contract.

Reply:
I broke it up into 7 separate specs and reduced code duplication via helper functions.